client_test.go

```diff
@@ -1,47 +1,184 @@
+	Convey("Given a valid session", t, func() {
+		mockRedisClient, mockClient := setUpMocks(*redis.NewStatusResult("success", nil), *redis.NewStringCmd())
 
-		mockClient := mock.ClientManagerMock{SetFunc: func(s *session.Session) error {
-			return nil
-		}}
+		s := &session.Session{
+			ID:           "1234",
+			Email:        "user@email.com",
+			Start:        time.Now(),
+			LastAccessed: time.Now(),
+		}
 
-		Convey("Where the session is valid", func() {
-			s := session.NewSession()
+		Convey("When cache attempts to store session", func() {
 			err := mockClient.Set(s)
 
-			Convey("Then the session should be Set in redis", func() {
-				So(mockClient.SetCalls(), ShouldHaveLength, 1)
+			Convey("Then session should be stored", func() {
 				So(err, ShouldBeNil)
+				So(mockRedisClient.SetCalls(), ShouldHaveLength, 1)
 			})
 		})
+	})
+
+	Convey("Given a valid session", t, func() {
+		mockRedisClient, mockClient := setUpMocks(*redis.NewStatusResult("fail", errors.New("failed to store session")), *redis.NewStringCmd())
+
+		s := &session.Session{
+			ID:           "1234",
+			Email:        "user@email.com",
+			Start:        time.Now(),
+			LastAccessed: time.Now(),
+		}
+
+		Convey("When cache attempts to store session", func() {
+			err := mockClient.Set(s)
 
+			Convey("Then session will not be stored", func() {
+				So(mockRedisClient.SetCalls(), ShouldHaveLength, 1)
+				So(err, ShouldNotBeEmpty)
+				So(err.Error(), ShouldEqual, "failed to store session")
+			})
+		})
 	})
 
-	Convey("Given redis client is set up", t, func() {
+	Convey("Given an invalid session", t, func() {
+		mockRedisClient, mockClient := setUpMocks(*redis.NewStatusCmd(), *redis.NewStringCmd())
 
-		mockClient := mock.ClientManagerMock{SetFunc: func(s *session.Session) error {
-			return errors.New("empty session provided")
-		}}
+		var s *session.Session = nil
 
-		Convey("Where the session is empty", func() {
-			var s *session.Session = nil
+		Convey("When cache attempts to store session", func() {
 			err := mockClient.Set(s)
 
-			Convey("Then session is not stored in redis", func() {
-				So(mockClient.SetCalls(), ShouldHaveLength, 1)
+			Convey("Then session will not be stored", func() {
+				So(mockRedisClient.SetCalls(), ShouldHaveLength, 0)
+				So(err, ShouldNotBeEmpty)
+				So(err.Error(), ShouldEqual, "session is empty")
+			})
+		})
+	})
+}
+
+func TestClient_GetByID(t *testing.T) {
+	Convey("Given a session ID", t, func() {
+		mockRedisClient, mockClient := setUpMocks(*redis.NewStatusCmd(), *redis.NewStringResult(string(resp), nil))
+
+		Convey("When cache attempts to get session", func() {
+			s, err := mockClient.GetByID("1234")
+			So(err, ShouldBeNil)
+
+			msg, err := mockRedisClient.Get(s.ID).Result()
+			So(err, ShouldBeNil)
+
+			var storedSession session.Session
+			err = json.Unmarshal([]byte(msg), &storedSession)
+			So(err, ShouldBeNil)
+
+			Convey("Then session is returned", func() {
+				So(storedSession.ID, ShouldEqual, s.ID)
```

Comment:
formatting?

Reply:
Turns out a lot was weird with how I wrote the `GetByID()` tests, gone through them and removed code that was duplicating the mocks behaviour.